codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/JsonSerializerGenerator.kt

```diff
@@ -82,14 +163,163 @@ class JsonSerializerGenerator(protocolConfig: ProtocolConfig) : StructuredDataSe
+                    serializeMember(context.member(member))
+                }
+            }
+        }
+        rust("#T(${context.objectName.borrowMut()}, ${context.localName});", structureSerializer)
+    }
+
+    private fun RustWriter.serializeMember(context: MemberContext) {
+        val target = model.expectShape(context.shape.target)
+        handleOptional(context) { inner ->
+            val key = inner.keyExpression
+            val value = inner.valueExpression.borrow()
+            when (target) {
+                is StringShape -> when (target.hasTrait<EnumTrait>()) {
+                    true -> context.writeValue(this, "string", key, "$value.as_str()")
+                    false -> context.writeValue(this, "string", key, value)
+                }
+                is BooleanShape -> context.writeValue(this, "boolean", key, value)
+                is NumberShape -> {
+                    val numberType = when (symbolProvider.toSymbol(target).rustType()) {
+                        is RustType.Float -> "Float"
+                        is RustType.Integer -> "NegInt"
+                        else -> throw IllegalStateException("unreachable")
+                    }
+                    context.writeInner(this, "number", key) {
+                        rustInline("#T::$numberType(*${inner.valueExpression})", smithyTypes.member("Number"))
+                    }
+                }
+                is BlobShape -> context.writeInner(this, "string_unchecked", key) {
+                    rustInline("&#T($value)", RuntimeType.Base64Encode(runtimeConfig))
+                }
+                is TimestampShape -> {
+                    val timestampFormat =
+                        httpIndex.determineTimestampFormat(context.shape, HttpBinding.Location.DOCUMENT, EPOCH_SECONDS)
+                    val timestampFormatType = RuntimeType.TimestampFormat(runtimeConfig, timestampFormat)
+                    context.writeInner(this, "instant", key) {
+                        rustInline("$value, #T", timestampFormatType)
+                    }
+                }
+                is CollectionShape -> jsonArrayWriter(inner) { arrayName ->
+                    serializeCollection(SimpleContext(arrayName, inner.valueExpression, target))
+                }
+                is MapShape -> jsonObjectWriter(inner) { objectName ->
+                    serializeMap(SimpleContext(objectName, inner.valueExpression, target))
+                }
+                is StructureShape -> jsonObjectWriter(inner) { objectName ->
+                    serializeStructure(StructContext(objectName, inner.valueExpression, target, symbolProvider))
                 }
+                is UnionShape -> jsonObjectWriter(inner) { objectName ->
+                    serializeUnion(SimpleContext(objectName, inner.valueExpression, target))
+                }
+                is DocumentShape -> {
+                    // TODO: Implement document shapes
```

Comment:
I would probably just leave this in the TODO or at least print a warning.

Reply:
The unit test will fail if I leave it as a Kotlin `TODO()` since I have a Document in the union, which is why I left it as the comment. I would hope our protocol tests would fail if I didn't address this comment :)